nexus/src/app/sagas/region_replacement_drive.rs

```diff
@@ -0,0 +1,1521 @@
+                            }
+
+                            _ => {
+                                ActionError::action_failed(
+                                    format!(
+                                        "unexpected failure during `disk_volume_status`: {e}",
+                                    )
+                                )
+                            }
+                        })?;
+
+                    // If the Volume is active, then reconciliation finished
+                    // successfully.
+                    //
+                    // There's a few reasons it may not be active yet:
+                    //
+                    // - Propolis could be shutting down, and tearing down the
+                    //   Upstairs in the process (which deactivates the Volume)
+                    //
+                    // - reconciliation could still be going on
+                    //
+                    // - reconciliation could have failed
+                    //
+                    // If it's not active, wait until the next invocation of
+                    // this saga to decide what to do next.
+
+                    result.into_inner().active
+                }
+
+                ReplaceResult::Missing => {
+                    // The disk's volume does not contain the region to be
+                    // replaced. This is an error!
+
+                    return Err(ActionError::action_failed(String::from(
+                        "saw ReplaceResult::Missing",
+                    )));
+                }
+            };
+
+            ExecuteResult { step_to_commit: Some(step), replacement_done }
+        }
+    };
+
+    Ok(result)
+}
+
+async fn srrd_drive_region_replacement_commit(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+
+    let opctx = crate::context::op_context_for_saga_action(
+        &sagactx,
+        &params.serialized_authn,
+    );
+
+    // If there was an executed step, record it!
+
+    let execute_result = sagactx.lookup::<ExecuteResult>("execute")?;
+
+    if let Some(step) = execute_result.step_to_commit {
+        info!(
+            log,
+            "committing step {}", step.step_time;
+            "region replacement id" => %params.request.id,
+        );
+
+        osagactx
+            .datastore()
+            .add_region_replacement_request_step(&opctx, step)
+            .await
+            .map_err(ActionError::action_failed)?;
+    } else {
+        info!(
+            log,
+            "no step to commit";
+            "region replacement id" => %params.request.id,
+        );
+    }
+
+    Ok(())
+}
+
+async fn srrd_drive_region_replacement_commit_undo(
+    _sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    // If this saga unwinds at the last step, do we have to remove the committed
```

Comment:
While I agree with the statement here, I'm concerned what happens if we didn't actually record the committed step in the DB in the last step based on your comment. The next time this saga is invoked, it should still be able to recover with only the prior step recorded.  If that is not the case, then we need to figure out how to change this saga to make that work.

Reply:
I think we're ok here, as long as the rule of

```
//! Note: _it's very important that only_ the Upstairs can make the
//! determination that a Volume is no longer degraded. Nexus should not be
//! assuming anything. This is the _golden rule_ that this saga must follow.
```

is followed. Nexus should not transition a region replacement request to ReplacementDone unless an Upstairs tells it that it's done.

If the final forward action fails to commit a step to the database, here are the cases where this saga could potentially repeat its action:

1. a propolis action was executed (read: a running propolis was sent a replace request)
2. a pantry action was executed (read: the volume was attached (activating in the background) to a pantry)

I think the saga unwinding then rerunning without knowledge of a previous step is the same as if the forward "execute" action reran. The drive saga wouldn't be correct unless each forward action was idempotent :) but I wrote it out here to convince myself:

# case 1

In the case of the next invocation of the drive saga choosing a propolis action:

- if the replace request is sent to the same propolis that originally received it, the upstairs would respond with `ReplaceResult::StartedAlready`. The drive saga would then consider the replacement not done and wait.

- if the replace request is sent to a different propolis, that propolis would have constructed the disk's volume with the replacement VCR, so the upstairs would respond with `ReplaceResult::VcrMatches`. The drive saga would then consider the replacement done only if propolis observed that the volume activated ok.

# case 2

In the case of the next invocation of the drive saga choosing a pantry action, Nexus first checks if the volume was already attached to the selected Pantry, and if so, will detach it before sending a "attach in the background with this job id" request.

- if Nexus chose the same Pantry as the original drive saga, this would cancel any existing reconciliation and start it up again from the beginning. This is ok - reconciliation can be interrupted at any time. If this repeatedly happened it would cause progress to be very slow, but progress would be made.

- if Nexus chose a different Pantry, the newly checked-out Volume would steal the activation from the original Pantry, cancelling the reconcilation only to start it up again on the different Pantry.

However, there is one small case where the saga could be smarter: pantry job IDs are unique per saga (and per step!), so in the case of the saga unwinding before it commits the step to the db we would lose it, but in the "forward action reran" case we would still have it. The `execute_pantry_drive_action` function could check if that job id exists, and if it does assume that it's in the rerun case - however I think that forward action reruns are rare enough that I don't want to bloat the code with more match statements. In the rare case that the reconciliation is interrupted, it'll resume without problems. I wrote a related comment in 8123053a8.

# also!

As well, I think both of these cases are equivalent to if Nexus chose to always attempt some sort of action, instead of choosing no-ops or waiting for operations driven by any previous steps to complete, aka if Nexus _always_ polled, instead of the behaviour it has now (wait or poll or receive push notifications). Polling all the time would be functionally correct but unnecessary (and in the case of crucible#1277, a problem!).

I committed this comment (more or less) in 42b23e762

Reply:
Awesome! Thank you for the details!